feat(group_theory/submonoid): add bundled submonoids and various lemmas

101damnations opened the pull request:
Not really sure when to put ```attribute [to_additive smul_something] something_pow``` after smul lemmas and when they do anything; there are some in ```group_theory.submonoid``` but not in every instance.

Docstrings: I'm not sure how clear 'natural powers' and 'natural multiples' is; I've written 'exponentiation/multiplication by naturals' where possible. The bundled and unbundled versions of lemmas might have non-identical docstrings sometimes as Kevin wrote some and I wrote some.

The majority of commits in here are irrelevant because I don't know how to branch properly; they're commits on files I deleted after branching as they're not to be included in this PR.

Edit: I don't know when I committed the extra lemmas because they were for other files, but the majority of the commits changing ```group_theory\submonoid.lean``` are: 
 - any of Kevin's except the last 3 before 30th August, 
 - ['using newer additive docstring format in submonoid'](https://github.com/leanprover-community/mathlib/pull/1623/commits/c0d5787cfb90b35557fd5e52d3e3972ba893ebb0), c0d5787
- anything after ['docstrings for group_theory.submonoid'](https://github.com/leanprover-community/mathlib/pull/1623/commits/6e9e045ee9d845384c80a402a453ccba9e24e797), 6e9e045

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.
[bundling submonoids on the #maths stream](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/bundling.20submonoids)

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

kbuzzard committed:
WIP -- removing  and everything is broken

kbuzzard committed:
WIP -- removing  and everything is broken

Commit:
test

Commit:
test

Commit:
test

Commit:
test

kbuzzard committed:
tidy

kbuzzard committed:
tidy

Commit:
fixed localization

Commit:
fixed localization

Commit:
starting on coset

Commit:
starting on coset

kbuzzard committed:
WIP

kbuzzard committed:
WIP

kbuzzard committed:
submonoid.lean now compiles but no to_additive stuff

kbuzzard committed:
submonoid.lean now compiles but no to_additive stuff

kbuzzard committed:
submonoid.lean compiles

kbuzzard committed:
submonoid.lean compiles

kbuzzard committed:
putting is_submonoid back

kbuzzard committed:
putting is_submonoid back

Commit:
docstrings

Commit:
docstrings

Commit:
terrible docstrings up to line 370

Commit:
terrible docstrings up to line 370

Commit:
finished docstrings

Commit:
finished docstrings

kbuzzard committed:
more to_additive stuff

kbuzzard committed:
more to_additive stuff

kbuzzard committed:
Merge branch 'bundled_submonoids' of github.com:leanprover-community/mathlib into bundled_submonoids

kbuzzard committed:
Merge branch 'bundled_submonoids' of github.com:leanprover-community/mathlib into bundled_submonoids

Commit:
WIP quotient monoids

Commit:
mathlib update

Commit:
quotient monoids WIP

Commit:
quotient_monoid w/o ideals.lean all compiles

Commit:
removing lemma

Commit:
adjunction

Commit:
some tidying

Commit:
remove pointless equiv

Commit:
completion compiles (very slowly)

Commit:
add lemma

101damnations committed:
tidying

101damnations committed:
more tidying

kbuzzard committed:
mul -> smul oops

kbuzzard committed:
might now compile

101damnations committed:
tidied! I think

kbuzzard committed:
Merge branch 'master' into bun_subs_with_new_toadditive

kbuzzard committed:
fix

kbuzzard committed:
Merge branch 'master' into bun_subs_with_new_toadditive

101damnations committed:
breaking/adding stuff & switching branch

101damnations committed:
add Inf relation

kbuzzard committed:
merge

kbuzzard committed:
fixing (trivial) ring conflict

kbuzzard committed:
removing sorrys

101damnations committed:
nearly updated quotient_monoid

101damnations committed:
update mathlib

101damnations committed:
updated quotient_monoid

101damnations committed:
resurrecting computability

101damnations committed:
tidied congruence.lean, added some docstrings

101damnations committed:
extending setoids instead, WIP

101damnations committed:
starting Galois insertion

kbuzzard committed:
a few more bits of to_additive and docs

kbuzzard committed:
no battery

kbuzzard committed:
up to line 800

101damnations committed:
congruence'll compile when data.setoid exists now

101damnations committed:
more updates modulo existence of data.setoid

101damnations committed:
rearranging stuff

101damnations committed:
docstrings

101damnations committed:
starting additive docstrings

101damnations committed:
using newer additive docstring format in submonoid

101damnations committed:
docstrings, tidying

101damnations committed:
fixes and to_additive stuff, all WIP

101damnations committed:
update mathlib

101damnations committed:
temporary congruence fixes

101damnations committed:
slightly better approach to kernels, general chaos

101damnations committed:
aahh

101damnations committed:
more mess

101damnations committed:
deleting doomed inductive congruence closure

101damnations committed:
many fixes and better char pred isoms

101damnations committed:
update mathlib & fix conflicts

101damnations committed:
docstrings for group_theory.submonoid

101damnations committed:
update mathlib

101damnations committed:
removing everything but bundled submonoids/lemmas

101damnations committed:
removing things etc

101damnations committed:
removing random empty folder

101damnations committed:
tidy

101damnations committed:
adding lemma back

101damnations committed:
tidying

ChrisHughes24 commented on src/algebra/group_power.lean:
```diff
@@ -340,6 +358,24 @@ theorem map_gsmul (a : α) (n : ℤ) : f (gsmul n a) = gsmul n (f a) :=
+namespace monoid_hom
+variables {β : Type v} [group α] [group β] (f : α →* β)
+
+theorem map_gpow (a : α) (n : ℤ) : f (a ^ n) = f a ^ n :=
```
These should all be `@[simp]`

ChrisHughes24 commented on src/algebra/commute.lean:
```diff
@@ -260,9 +260,19 @@ instance centralizer.is_submonoid : is_submonoid (centralizer a) :=
 { one_mem := commute.one_right a,
   mul_mem := λ _ _, commute.mul_right }
 
+def centralizer.submonoid : submonoid M :=
```
Could we have a docstring on `centralizer` pointing people towards these definitions. I also think the name is too long, but can't think of a better one.

kim-em replied:
Doc strings on every def, please!

ChrisHughes24 commented on src/algebra/pi_instances.lean:
```diff
@@ -356,6 +364,13 @@ instance [monoid α] [monoid β] [is_submonoid s] [is_submonoid t] :
   mul_mem := by intros; rw set.mem_prod at *; split; apply is_submonoid.mul_mem; tauto }
 
+@[to_additive add_submonoid]
+def submonoid [monoid α] [monoid β] (s : submonoid α) (t : submonoid β) :
```
This should be `submonoid.prod` so projection notation can be used.

kim-em commented on src/algebra/commute.lean:
```diff
@@ -310,9 +320,19 @@ instance centralizer.is_add_submonoid : is_add_submonoid (centralizer a) :=
 { zero_mem := commute.zero_right a,
   add_mem := λ _ _, commute.add_right }
 
+def centralizer.add_submonoid : add_submonoid A :=
```
Can this be achieved with `@[to_additive]` on `centralizer.submonoid`? (Also below?)

101damnations replied:
This would necessitate ```@[to_additive]``` on the definitions of ```commute``` and ```semiconj_by``` and a couple of associated lemmas, like ```semiconj_by.eq```, and am I right in thinking ```to_additive```'d versions of ```protected``` lemmas aren't also ```protected``` at the moment? I could just make an explicit ```protected lemma add_semiconj_by.eq``` if necessary though.

101damnations replied:
Wait, my bad. ```@to_additive```ing any amount of lemmas will not make this possible; the underlying set is still the multiplicative centralizer.

kim-em commented on src/algebra/pi_instances.lean:
```diff
@@ -241,6 +241,14 @@ lemma fst.is_monoid_hom [monoid α] [monoid β] : is_monoid_hom (prod.fst : α 
 lemma snd.is_monoid_hom [monoid α] [monoid β] : is_monoid_hom (prod.snd : α × β → β) :=
 { map_mul := λ _ _, rfl, map_one := rfl }
 
+@[to_additive prod.add_monoid_hom.fst]
```
Probably this is fine, but do you really need to give an explicit name for `to_additive` here?

jcommelin replied:
Probably yes, because otherwise you end up with `sum.add_monoid_hom.fst` which is not what you want.

kim-em replied:
Ah!

kim-em commented on src/data/equiv/algebra.lean:
```diff
@@ -505,12 +510,26 @@ variables [semiring α] [semiring β]
 { .. e.to_mul_equiv.to_monoid_hom, .. e.to_add_equiv.to_add_monoid_hom }
 
+/-- Reinterpret a ring equivalence as a monoid homomorphism. -/
+def to_monoid_hom (e : α ≃+* β) : α →* β := e.to_ring_hom.to_monoid_hom
```
I'm not sure this is a good idea; it saves a few keystrokes, but will surely require more lemmas to be written...

jcommelin replied:
@semorrison Would an `abbreviation` help here?

kim-em replied:
Okay.

kim-em commented on src/data/equiv/algebra.lean:
```diff
@@ -505,12 +510,26 @@ variables [semiring α] [semiring β]
+def to_monoid_hom (e : α ≃+* β) : α →* β := e.to_ring_hom.to_monoid_hom
+
+/-- Reinterpret a ring equivalence as an add_monoid homomorphism. -/
+def to_add_monoid_hom (e : α ≃+* β) : α →+ β := e.to_ring_hom.to_add_monoid_hom
```
Similarly here.

kim-em commented on src/data/equiv/algebra.lean:
```diff
@@ -505,12 +510,26 @@ variables [semiring α] [semiring β]
+  (h : R ≃* S) (H: ∀ x y : R, h (x + y) = h x + h y) : is_ring_hom h :=
+@ring_hom.is_ring_hom _ _ _ _ $ ring_hom.mk' h.to_monoid_hom H
+
+def of_mul_equiv {R : Type*} {S : Type*} [ring R] [ring S] (h : R ≃* S)
```
Possibly rename to `mul_equiv.to_semiring_hom`, so we can write `h.to_semiring_hom H`? Not certain this is better.

kim-em commented on src/group_theory/submonoid.lean:
```diff
@@ -65,66 +95,84 @@ lemma is_submonoid_Union_of_directed {ι : Type*} [hι : nonempty ι]
 
 section powers
 
+/-- The set of natural powers 1, x, x², ... of an element x of a monoid. -/
```
"natural number powers" does read better, here and below.

kim-em commented on src/group_theory/submonoid.lean:
```diff
@@ -65,66 +95,84 @@ lemma is_submonoid_Union_of_directed {ι : Type*} [hι : nonempty ι]
 | 0 := is_submonoid.one_mem s
 | (n + 1) := is_submonoid.mul_mem h is_submonoid.pow_mem
 
+/-- Add_submonoids are closed under multiplication by naturals. -/
```
This is getting picky, so please feel free to ignore, but I wouldn't capitalise "code". Generally, I think it's best if identifiers which are actual Lean code get surrounded by backticks. Sometimes it's harder to write a sensible sentence, but given that the tooltips do actually typeset the backticks, it can be helpful. This could be:

"An `add_submonoid` is closed under multiplication by ...."

kim-em commented on src/group_theory/submonoid.lean:
```diff
@@ -151,7 +201,8 @@ begin
 end
 
-@[to_additive]
+/-- The product of a finset of elements of a submonoid is an element of the submonoid. -/
```
It's not literally "a finset of elements of a submonoid" here, but a product _indexed by_ a `finset`.

101damnations committed:
responding to PR comments

101damnations committed:
change 2 defs to lemmas

ChrisHughes24 commented on src/algebra/commute.lean:
```diff
@@ -260,9 +261,22 @@ instance centralizer.is_submonoid : is_submonoid (centralizer a) :=
   mul_mem := λ _ _, commute.mul_right }
 
+/-- Centralizer of an element of a monoid `a` is the submonoid of elements that commute with `a`. -/
+@[to_additive hmm] def centralizer.submonoid : submonoid M :=
```
Is this a typo `@[to_additive hmm] `? I think maybe `submonoid.centralizer` is a better name, because then you can open the namespace and have a short name.

ChrisHughes24 commented on src/algebra/group_power.lean:
```diff
@@ -124,6 +124,24 @@ theorem map_smul (a : α) : ∀(n : ℕ), f (n • a) = n • (f a)
+namespace monoid_hom
+variables {β : Type v} [monoid α] [monoid β] (f : α →* β)
+
+theorem map_pow (a : α) : ∀(n : ℕ), f (a ^ n) = (f a) ^ n
```
Both of these should be `@[simp]`

101damnations committed:
@[simp] group_power.lean lemmas

101damnations committed:
responding to commute.lean comments

101damnations committed:
Remove unnecessary add_semiconj_by.eq

101damnations committed:
Change prod.submonoid to submonoid.prod

101damnations committed:
replacing a / at the end of docstring

Sorry - don't make commits on your phone when your laptop's playing up :/

101damnations committed:
removing some not very useful to_additives

101damnations committed:
Merge branch 'master' into bundled_submonoids_etc

101damnations committed:
fix pi_instances namespaces

101damnations committed:
remove unnecessary prefix

101damnations committed:
change extensionality to ext

not sure this is necessary because surely merging will change this automatically, but Travis told me to, and I really want it to compile, and I am not at my laptop

mergify[bot] committed:
Merge branch 'master' into bundled_submonoids_etc
